hack/test

```diff
@@ -1,20 +1,90 @@
 #!/usr/bin/env bash
 
-set -eu -o pipefail -x
+. $(dirname $0)/util
+set -eu -o pipefail
 
+: ${TEST_INTEGRATION=}
+: ${TEST_GATEWAY=}
+: ${TEST_DOCKERFILE=}
+: ${CONTINUOUS_INTEGRATION=}
+
+progressFlag=""
+if [ "$CONTINUOUS_INTEGRATION" == "true" ]; then progressFlag="--progress=plain"; fi
+
+if [ "$#" == 0 ]; then TEST_INTEGRATION=1; fi
+
+while test $# -gt 0
+  do
+    case "$1" in
+        gateway)
+          TEST_GATEWAY=1
+        ;;
+        dockerfile)
+          TEST_DOCKERFILE=1
+        ;;
+        integration)
+          TEST_INTEGRATION=1
+        ;;
+        *)
+          echo "unknown arg $1"
+        ;;
+    esac
+    shift
+done
+
+
+iid="buildkit-tests"
 iidfile=$(mktemp -t docker-iidfile.XXXXXXXXXX)
+set -x
+
+case $buildmode in
+"buildkit")
+  tmpfile=$(mktemp -t docker-iidfile.XXXXXXXXXX)
+  buildctl build $progressFlag --frontend=dockerfile.v0 --local context=. --local dockerfile=. \
+    --frontend-opt target=integration-tests \
+    --frontend-opt filename=./hack/dockerfiles/test.buildkit.Dockerfile \
+    --exporter=docker --exporter-opt name=$iid --exporter-opt output=$tmpfile
+  docker load -i $tmpfile
+  rm $tmpfile
+  ;;
+*)
+  case $buildmode in
```

Comment:
I don't understand this `case $buildmode` within the `*)` of a `case $buildmode`. Bad refactoring or am I missing something?

Reply:
the `*` case is longer than inner switch